### PR TITLE
etcdserver: use context for Renew

### LINKED ISSUE #6922: client: user may freeze on KeepAlive
**etcd client version:** `3.1.0-rc.0` (git hash: 83347907)

**Issue can be reproduced in following way:**
1) Create an etcdv3 client.
2) Make `lessor.recvKeepAliveLoop()` fail with any server-side error (we got `"lease: unknown error(not a primary lessor )"`).
3) Create a new lease.
4) Try to keep it alive with `KeepAlive`.

**Expected behavior:**
We expected client to recover from `recvKeepAliveLoop` error after some time and continue to process leases. Alternatively we expected `KeepAlive` to return error immediately due to broken `recvKeepAliveLoop`, so we can restart client manually, at least.

**Observed behavior:**
User of `KeepAlive` freezes because returned channel is never closed, after `keepAliveCtxCloser` exits on [`<-l.donec`](https://github.com/coreos/etcd/blob/83347907774bf36cbb261c594a32fd7b0f5dd9f6/clientv3/lease.go#L275) case in select due to `recvKeepAliveLoop` [signal](https://github.com/coreos/etcd/blob/83347907774bf36cbb261c594a32fd7b0f5dd9f6/clientv3/lease.go#L333). User code assumes that keep alive still works and expects responses or close of the channel, but none of these events happens.

**Side notes**
We used `concurrency.Session` to control the lease, and in this case it leads to [goroutine leak](https://github.com/coreos/etcd/blob/83347907774bf36cbb261c594a32fd7b0f5dd9f6/clientv3/concurrency/session.go#L59) and freeze of `session.Close` (or more precisely [session.Orphan](https://github.com/coreos/etcd/blob/83347907774bf36cbb261c594a32fd7b0f5dd9f6/clientv3/concurrency/session.go#L84)).

### PR DESCRIPTION
**Pull request by heyitsanthony**
Renew would retry a few times before returning a not primary error that
the client should never see. Instead, use proper timeouts and
then return a request timeout error on failure.

Fixes #6922

### CODE REVIEW

**Comment by gyuho** [comment]
lgtm. Defer to @xiang90

File: clientv3/integration/lease_test.go [hunk 1]
```diff
@@ -510,3 +510,44 @@ func TestLeaseTimeToLive(t *testing.T) {
...
+	defer kcancel()
+	ka, err := cli.KeepAlive(kctx, r.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// consume first keepalive so next message sends when cluster is down
+	<-ka
+
+	// force keepalive stream message to timeout
+	clus.Members[1].Stop(t)
```
Comment (xiang90) [inline clientv3/integration/lease_test.go]: how do we ensure member[0] is not the leader?

File: lease/leasehttp/http.go [hunk 2]
```diff
@@ -125,15 +124,22 @@ func (h *leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 	}
 
-	cc := &http.Client{Transport: rt, Timeout: timeout}
-	resp, err := cc.Post(url, "application/protobuf", bytes.NewReader(lreq))
+	cc := &http.Client{Transport: rt}
+	req, err := http.NewRequest("POST", url, bytes.NewReader(lreq))
+	if err != nil {
+		return -1, err
+	}
+	req = req.WithContext(ctx)
```
Comment (xiang90) [inline lease/leasehttp/http.go]: does go16 support this?

File: clientv3/integration/lease_test.go (hunk 1 above)
Comment (xiang90) [inline clientv3/integration/lease_test.go]: leader will stepdown after X seconds I believe, not immediately.

Comment (heyitsanthony) [inline clientv3/integration/lease_test.go]: OK, but for the duration it's not going to be able to expire any keys. The test just needs to get a renew failure + lease extension. Maybe this test should be renamed `TestLeaseRenewLostQuorum`?

File: lease/leasehttp/http.go (hunk 2 above)
Comment (heyitsanthony) [inline lease/leasehttp/http.go]: ugh 3.0 backporting, I'll fix this to do it without it

**Comment by xiang90** [comment]
lgtm

**Comment by heyitsanthony** [comment]
I assume this is to be backported if 1.6 compatibility is important

**Pull request merged**